### ISSUE TITLE
The recent commit in ifvc.c to improve strict aliasing is not compiling for eg freeBSD 11.0

### ISSUE THREAD

**Issue by bringha1** [pali/igmpproxy#24]
The recent commit with the change of ifvc.c does not compile for BSD variants; eg line 123
`mask = s_addr_from_sockaddr(&IfReq.ifr_netmask);` leads to compiler error as being linux specific; struct ifreq in eg freebsd 11.1 does not contain an element ifr_netmask; 

/usr/include/net/if.h
(....)
` /*
 * Interface request structure used for socket
 * ioctl's.  All interface ioctl's must have parameter
 * definitions which begin with ifr_name.  The
 * remainder may be interface specific.
 */
struct  ifreq {
        char    ifr_name[IFNAMSIZ];             /* if name, e.g. "en0" */
        union {
                struct  sockaddr ifru_addr;
                struct  sockaddr ifru_dstaddr;
                struct  sockaddr ifru_broadaddr;
                struct  ifreq_buffer ifru_buffer;
                short   ifru_flags[2];
                short   ifru_index;
                int     ifru_jid;
                int     ifru_metric;
                int     ifru_mtu;
                int     ifru_phys;
                int     ifru_media;
                caddr_t ifru_data;
                int     ifru_cap[2];
                u_int   ifru_fib;
                u_char  ifru_vlan_pcp;
        } ifr_ifru;
#define ifr_addr        ifr_ifru.ifru_addr      /* address */
#define ifr_dstaddr     ifr_ifru.ifru_dstaddr   /* other end of p-to-p link */
#define ifr_broadaddr   ifr_ifru.ifru_broadaddr /* broadcast address */
#define ifr_buffer      ifr_ifru.ifru_buffer    /* user supplied buffer with its length */
#define ifr_flags       ifr_ifru.ifru_flags[0]  /* flags (low 16 bits) */
#define ifr_flagshigh   ifr_ifru.ifru_flags[1]  /* flags (high 16 bits) */
#define ifr_jid         ifr_ifru.ifru_jid       /* jail/vnet */
#define ifr_metric      ifr_ifru.ifru_metric    /* metric */
#define ifr_mtu         ifr_ifru.ifru_mtu       /* mtu */
#define ifr_phys        ifr_ifru.ifru_phys      /* physical wire */
#define ifr_media       ifr_ifru.ifru_media     /* physical media */
#define ifr_data        ifr_ifru.ifru_data      /* for use by interface */
#define ifr_reqcap      ifr_ifru.ifru_cap[0]    /* requested capabilities */
#define ifr_curcap      ifr_ifru.ifru_cap[1]    /* current capabilities */
#define ifr_index       ifr_ifru.ifru_index     /* interface index */
#define ifr_fib         ifr_ifru.ifru_fib       /* interface fib */
#define ifr_vlan_pcp    ifr_ifru.ifru_vlan_pcp  /* VLAN priority */
};
`

**Comment by pali** [comment]
Hi! Can you give me some pointers to documentation for SIOCGIFNETMASK ioctl?

**Comment by bringha1** [comment]
What precisely do you need? I think all the usual suspects like intro(4) and ioctl(2) from the BSD space you know ....

**Comment by fichtner** [comment]
Nevertheless, FreeBSD port for version 0.2 was submitted with last minute reverts that broke the build. See attached patch for successful building. Tested ok on OPNsense.

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=224543

**Comment by pali** [comment]
@bringha1 I mean exact API of SIOCGIFNETMASK ioctl for FreeBSD (what is input structure, what is output).

For example that ioctl on Linux is documented in netdevice manpage where is written:
https://linux.die.net/man/7/netdevice
```
SIOCGIFNETMASK, SIOCSIFNETMASK
Get or set the network mask for a device using ifr_netmask. For compatibility, only AF_INET addresses are accepted or returned. Setting the network mask is a privileged operation.
```

**Comment by bringha1** [comment]
For example here:
[https://www.freebsd.org/cgi/man.cgi?query=ifnet&sektion=9](url)

or

[https://www.freebsd.org/cgi/man.cgi?query=inet&sektion=4&apropos=0&manpath=FreeBSD+8.0-RELEASE](url)

**Comment by pali** [comment]
I already found them, but there is nothing useful. First link does not contain anything about SIOCGIFNETMASK and second link specify just `same form as the SIOCIFADDR	command	(see intro(4))`. But `intro(4)` link again does not specify anything about SIOCIFADDR.

As `ifr_ifru` is just union of more `struct sockaddr` structures, we could use any sockaddr, all should contain same data. But I wanted to implement it in way how documentation suggest. On  Linux documentation is written to use `ifr_netmask`, but I cannot find anything for BSD...

**Comment by bringha1** [comment]
Well …

another thing what I found is:

[https://www.freebsd.org/cgi/man.cgi?query=netintro&sektion=4&apropos=0&manpath=FreeBSD+5.3-RELEASE+and+Ports](url)

and also some more general thing:

[https://www.freebsd.org/doc/en/books/developers-handbook/index.html](url) look into chapter 7

This is what I usually use: Chapter 4 of the man pages and the developer guides ….

Br

**Comment by pali** [comment]
Nothing in last two links too.

**Comment by pali** [comment]
Ok, no help, so I looked into freebsd kernel sources and they fill `ifr_addr` member for SIOCGIFNETMASK ioctl:

https://github.com/freebsd/freebsd/blob/c92659280baa4f5efb81c89074f1cea55005a9bd/sys/netinet/in.c#L229-L230
```
	struct ifreq *ifr = (struct ifreq *)data;
	struct sockaddr_in *addr = (struct sockaddr_in *)&ifr->ifr_addr;
```

https://github.com/freebsd/freebsd/blob/c92659280baa4f5efb81c89074f1cea55005a9bd/sys/netinet/in.c#L324-L326
```
	case SIOCGIFNETMASK:
		*addr = ia->ia_sockmask;
		break;
```

I will update igmpproxy code to use `ifr_addr` member.

**Comment by pali** [comment]
Fixed now in new release 0.2.1: https://github.com/pali/igmpproxy/releases/tag/0.2.1
